### PR TITLE
SAML SSO Authentication System

### PR DESCRIPTION
**Pull request by rishirajbansal**
Hi Louis,

Codebase for SAML SSO Authentication system is committed.

It consists of 2 projects:

- Django Admin
- Pulumi

Both of the projects need to build separately. 

Reason to create separate project for Pulumi is due to the Python version conflicts between Pulumi and warrant library. **Pulumi prerequisite version for Python is 3.6 or above while for warrant it is 3.5.** That means warrant will not work if Python 3.6 version is used. Warrant is not supporting version above 3.5. I tried to make it as a single project, attempt to use various Python versions of 3.6.x but Warrant is not supporting it. 

Details on project installment and deployment are drafted in the document and shared already.

Please let me know if you see any conflicts.

Thanks,
Rishi

### CODE REVIEW

File: authenticator/application/generic/init/initConfigurator.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import logging
+from application.generic.base.exception.genericExceptions import ApplicationException
+
+
+class InitConfigurator:
```
Comment (LouisAmon) [inline authenticator/application/generic/init/initConfigurator.py]: What is this for ?

File: authenticator/application/generic/loggerManager/logManager.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import os
+
+
+class LogManager:
```
Comment (LouisAmon) [inline authenticator/application/generic/loggerManager/logManager.py]: Why would you define your own log manager in the context of creating a simple Cognito authentication backend ?

Can't you let people choose how they treat their logs ?

File: authenticator/application/generic/management/commands/createSuperCogUser.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+from application.authenticate.integration.cognito.facade.userManagement import CognitoUserMgmt
+from application.generic.base.constants.exceptionConstants import ExceptionConstants
+from application.generic.base.exception.businessExceptions import CognitoUserMgmtException
+from application.generic.base.exception.exceptionUtility import ExceptionUtility
+
+
+class Command(createsuperuser.Command):
+    logger = logging.getLogger(__name__)
+
+    help = 'Create a superuser in Djnago Admin along with Cogntio User Pool'
```
Comment (LouisAmon) [inline authenticator/application/generic/management/commands/createSuperCogUser.py]: Typo in help

Why do we need to create an additional command ?

If Cognito is set as the default backend, then wouldn't the regular command do the trick ?

File: authenticator/application/generic/base/constants/businessConstants.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+
+
+class BusinessConstants:
+    RESPONSE_SUCCESS = "success"
+    RESPONSE_FALSE = "false"
+    RESPONSE_ERROR = "error"
```
Comment (LouisAmon) [inline authenticator/application/generic/base/constants/businessConstants.py]: This needs to go

File: authenticator/application/generic/base/constants/genericConstants.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+
+
+class GenericConstants:
+
+    EMPTY_STRING = ""
+    STRING_TRUE = "true"
+    STRING_FALSE = "false"
+    STRING_YES = "Y"
+    STRING_NO = "N"
```
Comment (LouisAmon) [inline authenticator/application/generic/base/constants/genericConstants.py]: Is this Cognito standard return values ?

The naming and docstring don't indicate anything such.

It seems very redundant

File: authenticator/application/generic/base/exception/genericExceptions.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+from application.generic.base.exception.exceptionDetail import *
+
+
+class ApplicationException(Exception):
```
Comment (LouisAmon) [inline authenticator/application/generic/base/exception/genericExceptions.py]: Needs to go

File: authenticator/application/generic/utilities/utility.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+class Utility:
+
+    current_time_in_milli = lambda: int(round(time.time() * 1000))
+
+    @staticmethod
+    def safe_trim(string):
+        if string is None or string.strip() == '':
+            return ''
+        else:
+            return string.strip()
```
Comment (LouisAmon) [inline authenticator/application/generic/utilities/utility.py]: I'm against creating a `utility` module when it is avoidable as it obscures the code with functions of which you don't necessarily understand the purpose in context.

Can the output strings of Cognito really be `   ` (string with spaces) ?

Either way, this module needs to go and the logic should be applied where it is needed (maybe not all values can return strings with spaces)

Also, this kind of stuff (odd empty string with spaces) should typically be handled in test cases.

File: authenticator/application/authenticate/integration/cognito/backends/emailPwdBackend.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+from application.authenticate.integration.cognito.facade.userManagement import CognitoUserMgmt
+from application.generic.base.exception.exceptionUtility import ExceptionUtility
+from application.generic.base.constants.genericConstants import *
+from application.generic.base.constants.exceptionConstants import *
+from application.generic.base.exception.genericExceptions import *
+from application.generic.base.exception.businessExceptions import *
+
+
+class CognitoAuthBackend():
```
[the remark this answers is not in the document]
Comment (rishirajbansal) [reply authenticator/application/authenticate/integration/cognito/backends/emailPwdBackend.py]: The Django Version used in this project is v2.2 and it does not suppot BaseBackend class.
Please see:
https://docs.djangoproject.com/en/2.2/topics/auth/customizing/#authentication-backends
 Its need to add in settings files like this:

AUTHENTICATION_BACKENDS = [
    'application.authenticate.integration.cognito.backends.samlBackend.CognitoAuthBackend',
    'application.authenticate.integration.cognito.backends.emailPwdBackend.CognitoAuthBackend',
    # 'django.contrib.auth.backends.ModelBackend',

]

File: authenticator/application/authenticate/integration/cognito/facade/facade.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        refresh_token = cognito.refresh_token
+
+        user = cognito.get_user(attr_map={"email": "email"})
+        user_emnail = user._data['email']
+
+        # Delete User
+        # cognito.authenticate(password='Nauge009#')
+        # cognito.delete_user()
+
+        self.logger.debug("Cognito:")
```
[the remark this answers is not in the document]
Comment (rishirajbansal) [reply authenticator/application/authenticate/integration/cognito/facade/facade.py]: Sorry about that, I missed to remove.

File: authenticator/application/authenticate/integration/cognito/backends/emailPwdBackend.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (LouisAmon) [reply authenticator/application/authenticate/integration/cognito/backends/emailPwdBackend.py]: Why are you using Django 2.2 ?

Current version is 3.0

[the remark this answers is not in the document]
Comment (rishirajbansal) [reply authenticator/application/authenticate/integration/cognito/backends/emailPwdBackend.py]: To avoid any conflicting issues with Python v 3.5

**Pull request merged**